specification/botservice/resource-manager/Microsoft.BotService/preview/2017-12-01/botservice.json

```diff
@@ -796,6 +796,423 @@
           "nextLinkName": "nextLink"
         }
       }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.BotService/BotServices/listServiceProviders": {
+      "post": {
+        "tags": [
+          "ListServiceProviders"
+        ],
+        "description": "Lists the available Service Providers for creating Connection Settings",
+        "operationId": "BotConnections_ListServiceProviders",
+        "x-ms-examples": {
+          "List Auth Service Providers": {
+            "$ref": "./examples/ListServiceProviders.json"
+          }
+        },
+        "parameters": [
+          {
+            "name": "resourceGroupName",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "description": "The name of the resource group within the user's subscription."
+          },
+          {
+            "$ref": "#/parameters/apiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/subscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "If resource is retrieved successfully, the service should return 200 (OK).",
+            "schema": {
+              "$ref": "#/definitions/ServiceProviderResponseList"
+            }
+          },
+          "default": {
+            "description": "Error response describing why the operation failed",
+            "schema": {
+              "$ref": "#/definitions/Error"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.BotService/botServices/{resourceName}/Connections/{connectionName}/listWithSecrets": {
+      "post": {
+        "tags": [
+          "BotConnection"
+        ],
+        "x-ms-examples": {
+          "Update Connection Setting": {
+            "$ref": "./examples/GetConnection.json"
+          }
+        },
+        "description": "Get a Connection Setting registration for a Bot Service",
+        "operationId": "BotConnection_ListWithSecrets",
+        "parameters": [
+          {
+            "name": "resourceGroupName",

```

Comment:
An empty `resourceGroupName` being sent over the wire can result in really obfuscated errors being returned to the client. Please add some validation to the definition of `resourceGroupName` to prevent this.

Reply:
added validation similar to resource name